### PR TITLE
[LOOP-4096] adding programmed units to cached insulin delivery object

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-4096

### CODE REVIEW

File: LoopKit.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -5911,10 +5912,11 @@
 		C168CA77280DD00F002BD2A7 /* Model.xcdatamodeld */ = {
 			isa = XCVersionGroup;
 			children = (
+				B44E84C32C21B164000B7C76 /* Modelv5.xcdatamodel */,
 				C168CA78280DD00F002BD2A7 /* Model.xcdatamodel */,
 				C168CA79280DD00F002BD2A7 /* Modelv4.xcdatamodel */,
 			);
-			currentVersion = C168CA79280DD00F002BD2A7 /* Modelv4.xcdatamodel */;
+			currentVersion = B44E84C32C21B164000B7C76 /* Modelv5.xcdatamodel */;
```
Comment (ps2) [inline LoopKit.xcodeproj/project.pbxproj]: Adding a new optional field should not require a model rev. As long as the migration tests pass, you should be able to just modify the existing v4 model.

File: LoopKitHostedTests/InsulinDeliveryStoreTests.swift [hunk 2]
```diff
@@ -208,7 +208,7 @@ class InsulinDeliveryStoreTests: InsulinDeliveryStoreTestsBase {
         XCTAssertEqual(entries[0].endDate, self.entry1.endDate)
         XCTAssertEqual(entries[0].value, 0.015)
```
Comment (ps2) [inline LoopKitHostedTests/InsulinDeliveryStoreTests.swift:209]: Adding a new `deliveredUnits` field to `CachedInsulinDeliveryObject` is probably the right approach for representing an interrupted bolus. Also, for basal doses this is probably an improvement over the old method of stashing a `DoseEntry`'s deliveredUnits into `value` and changing the `unit` to `.units`. So to be complete with this change, for basal doses, I think the `value` field here should now show the original programmed rate of 1.8, and the `unit` field should be `unitsPerHour`.  It also makes the `programmedTempBasalRate` field on `CachedInsulinDeliveryObject` redundant. I think that can also be dropped without needing a new model version.

**Review by ps2 — COMMENTED** [review]
I think this is a good approach, and the code so far looks good. I think there are a couple additional step needed. 1) see my comment about updating `value` and `units` of basal doses. 2) We should also extend storage of `programmedUnits` to HealthKit. So we may need another metadata key like `MetadataKeyProgrammedTempBasalRate`, but for boluses.

File: LoopKit.xcodeproj/project.pbxproj (hunk 1 above)
Comment (nhamming) [reply LoopKit.xcodeproj/project.pbxproj]: I added migration to change `value` -> `deliveredUnits` to make this more explicit. Should `value` be kept?

File: LoopKitHostedTests/InsulinDeliveryStoreTests.swift (hunk 2 above)
Comment (nhamming) [reply LoopKitHostedTests/InsulinDeliveryStoreTests.swift:209]: note that the new field is `programmedUnits`. The original `value` in the `CachedInsulinDeliveryObject` was just renamed to `deliveredUnits`.

File: LoopKitTests/DoseStoreTests.swift [hunk 3]
```diff
@@ -927,8 +927,8 @@ class DoseStoreTests: PersistenceControllerTestCase {
         XCTAssertEqual(doseEntries[1].type, .basal)
         XCTAssertEqual(doseEntries[1].startDate, f("2018-12-12 18:05:00 +0000"))
         XCTAssertEqual(doseEntries[1].endDate, f("2018-12-12 18:25:00 +0000"))
-        XCTAssertEqual(doseEntries[1].value, 0.3)
-        XCTAssertNil(doseEntries[1].deliveredUnits)
+        XCTAssertEqual(doseEntries[1].value, 0.85)
```
Comment (ps2) [inline LoopKitTests/DoseStoreTests.swift:946]: Maybe there should be a unit assertion here too:
```
XCTAssertEqual(doseEntries[1].unit, .unitsPerHour)
```

Comment (ps2) [reply LoopKitTests/DoseStoreTests.swift:946]: And on the other similar assertions for `value` that are missing the unit.

**Review by ps2 — APPROVED** [review]
LGTM!

**Comment by nhamming** [comment]
@ps2 could to do one more review. @SwiftlyNoah added his update to the model to support favorite foods.

**Review by ps2 — APPROVED** [review]
LGTM!

**Pull request merged**